### PR TITLE
Optimize the box rearrange create algorithm

### LINKED ISSUE #80: PIO_initdecomp might take a long time if box rearranger is used in PIO2
This issue has been confirmed by GPTL timing information of some high-resolution ACME F cases run on supercomputers.

[Cori@NERSC, 17408 MPI tasks]
PIO1: 1 min 25 sec for 76 PIO_initdecomp calls, 1.12 sec per call
PIO2: 17 min 34 sec for 76 PIO_initdecomp calls, 13.87 sec per call

[Titan@OLCF, 43200 MPI tasks]
PIO1: 2 min 45 sec for 76 PIO_initdecomp calls, 2.17 sec per call
PIO2: 67 min 21 sec for 76 PIO_initdecomp calls, 53.17 sec per call

When box rearranger is used, PIO_initdecomp calls box_rearrange_create. Most of the run time is spent in pio_swapm (called by box_rearrange_create). box_rearrange_create is implemented differently in PIO1, which uses "gather and then broadcast" instead of calling pio_swapm.

PIO2 should try to reduce the run time of box_rearrange_create (comparable to PIO1) with a different implementation.

### PR DESCRIPTION
**Pull request by jayeshkrishna**
Optimize the box rearrange create algorithm by reducing the
number of collectives used to exchange information between
compute and I/O processes.

Also see PR #109 that optimized the same algorithm.

Fixes #80

### CODE REVIEW

**Comment by jayeshkrishna** [comment]
With the change the performance of high res F case on Cori with no restarts and hourly outputs improves by 2X (~500 MB/s to ~1GB/s).

**Comment by dqwu** [comment]
For the high res F case on Cori (with no restarts):
```
Case run time: 20 min 17 sec ==> 14 min 1 sec
a:CAM_run4: 46.656 sec ==> 20.786 sec
a:wshist: 32.351 sec ==> 6.438 sec
a:dump_field: 11.467 sec ==> 1.786 sec
PIO:PIOc_initdecomp: 122.801 sec ==> 25.118 sec
a:PIO:initdecomp_dof: 19.886 sec ==> 3.151 sec
```

File: src/clib/pio_rearrange.c [hunk 1]
```diff
@@ -1229,6 +1229,276 @@ int box_rearrange_create(iosystem_desc_t *ios, int maplen, const PIO_Offset *com
...
+    {
+        if (!(dest_ioproc = malloc(maplen * sizeof(int))))
+            return pio_err(ios, NULL, PIO_ENOMEM, __FILE__, __LINE__);
+
+        if (!(dest_ioindex = malloc(maplen * sizeof(PIO_Offset))))
+            return pio_err(ios, NULL, PIO_ENOMEM, __FILE__, __LINE__);
+    }
+
+    /* Initialize the sc_info send and recv messages */
+    for(int i=0; i<sc_info_msg_sz; i++)
```
Comment (dqwu) [inline src/clib/pio_rearrange.c]: Is it better to use memset here?

File: src/clib/pio_rearrange.c [hunk 2]
```diff
@@ -1229,6 +1229,276 @@ int box_rearrange_create(iosystem_desc_t *ios, int maplen, const PIO_Offset *com
...
+    {
+        sc_info_msg_send[i] = 0;
+    }
+
+    /* The sc_info_msg_recv[i * sc_msg_info_sz] contains the sc_info from
+     * iorank i (the union rank for iorank i is ios->ioranks[i]). Each
+     * sc_info message contains [iomaplen, start_for_all_dims, count_for_all_dims]
+     */
+    for(int i=0; i<ios->num_iotasks * sc_info_msg_sz; i++)
```
Comment (dqwu) [inline src/clib/pio_rearrange.c]: Same as above, memset can be used here.

Comment (jayeshkrishna) [reply src/clib/pio_rearrange.c]: We could perform such optimizations in another PR, if needed (most likely the compiler will do it for us).

**Comment by ndkeen** [comment]
Clarifying that this only affects the code when using PIO2?

**Comment by jayeshkrishna** [comment]
Yes, these optimizations are in the E3SM fork (master) of PIO2.

**Pull request merged**